### DISCUSSION TITLE
Tinkerboard S with rtl8812au lost packages. Can not get stable video.

### DISCUSSION THREAD

**Discussion by marstears** [svpcom/wfb-ng#129] [Questions]
**Hardware：ASUS Tinkerboard S (rk3288)
Wifi module：ASUS AC56 USB（rtl8812au）
OS：TinkerOS with linux kernal 4.4
Wifi Driver：patched driver v5.6.4.2
Camera: IMX219
Both RX and TX have the same environment and tested indoor.**

**wifibroadcast.cfg settings for video:**
```
[common]
wifi_channel = 165
tx_sel_delta = 3 
wifi_region = 'GY' 
debug = False
radio_mtu = 1446
mavlink_agg_timeout = 0.1
#wifi_txpower = None
wifi_txpower = 58        # for 8812au

[drone_video]
keypair = 'drone.key'
stats_port = None
stream = 3
fec_k = 8 
fec_n = 12 
peer = 'listen://0.0.0.0:5600' 
bandwidth = 20 
short_gi = False
stbc = 1
ldpc = 0
mcs_index = 1

[gs_video]
keypair = 'gs.key' 
stats_port = 8002 
stream = 3 
fec_k = 8 
fec_n = 12 
peer = 'connect://127.0.0.1:5600' 
bandwidth = 20

```
**TX video:**
`gst-launch-1.0 rkcamsrc device=/dev/video0 io-mode=4 isp-mode=2A tuning-xml-path=/etc/cam_iq/IMX219.xml ! queue ! video/x-raw,format=NV12,width=1280,height=720,framerate=24/1 ! queue ! mpph264enc ! 'video/x-h264,stream-format=(string)byte-stream' ! rtph264pay ! queue ! udpsink host=127.0.0.1 port=5600 -v`

**RX video played with QgroundControl.**

**wfb-cli gs:**
![微信截图_20210402181005](https://user-images.githubusercontent.com/28802367/113407968-ffe3ab00-93e0-11eb-9a7c-4be10df8c1f1.png)

**ping from tx to rx tunnel lost about 10%-20% packages.**
![image](https://user-images.githubusercontent.com/28802367/113408281-a3cd5680-93e1-11eb-91fb-13f6a6ad1f03.png)

**Video trans with test script.**
![image](https://user-images.githubusercontent.com/28802367/113408785-982e5f80-93e2-11eb-80c9-0ebb1458fe91.png)

**I have perfect video trans with LAN, but because of package lost I can not even see any continuous image through wifibroadcast. I tried to change settings such like wifi_channel, msc_index and bandwidth got no changes. And I have tested 4 same ASUS AC56 wifi modules with exactly same results. Please help me.**

**Comment by svpcom** [comment]
1. Check that your video **bitrate** < 7mbit/s.  Instead of real cam use `gst-launch-1.0 videotestsrc ! videoconvert ! video/x-raw,format=NV12,framerate=30/1,width=1280,height=720 ! x264enc bitrate=4000 tune=zerolatency ! rtph264pay config-interval=1 ! udpsink host=127.0.0.1 port=5600`
2. Check that no NetworkManager and/or wpa_supplicant are on used wifi adapters.
3. Try other wifi channel

**Reply by marstears** [reply]
Thank you for replay！
I tried your method, but no change.
I already add my wifi adapter to /etc/NetworkManager/NetworkManager.conf. Change channel only affect rssi.
When I stop video streaming and ping from tx to rx from tunnel still have 10%-20% package lost. And in tunnel section with wfb-cli “Link Lost” and “package lost” appear randomly alternately. So I don't think this problem has anything to do with image transmission.
![image](https://user-images.githubusercontent.com/28802367/113705301-d80d8380-970f-11eb-8764-8188ca6122ca.png)

BTW, I find my wifi adapter @drone have IP address when wifibroadcast is stoped.
But no IP address @gs. I don't know if it will affect my situation.
**@drone**
![image](https://user-images.githubusercontent.com/28802367/113706641-836b0800-9711-11eb-971b-28e323e59a8c.png)
**@gs**
![image](https://user-images.githubusercontent.com/28802367/113706688-9087f700-9711-11eb-987d-4a4005291749.png)

**Reply by svpcom** [reply]
Channel change should not affect RSSI. Check that your WIFI channel is not used by other stations. Also check that NetworkManager ignores it (nmcli should show card in "unmanaged" state).

**Reply by svpcom** [reply]
If someone set ip to card then you should find and disable it

**Reply by marstears** [reply]
I changed channel to 165 which is no used by any station in our office (I use wifi scan software to make sure it).
And this is my nmcli result.
![image](https://user-images.githubusercontent.com/28802367/113707338-58cd7f00-9712-11eb-8036-3723a21b8983.png)

**Reply by svpcom** [reply]
Try to insert both cards into GS. And run:
```
wfb-server gs wlanA
wfb-server drone wlanB
```
then run:

```
gst-launch-1.0 videotestsrc ! videoconvert ! video/x-raw,format=NV12,framerate=30/1,width=1280,height=720 ! x264enc bitrate=4000 tune=zerolatency ! rtph264pay config-interval=1 ! udpsink host=127.0.0.1 port=5602

gst-launch-1.0 udpsrc port=5600 caps='application/x-rtp, media=(string)video, clock-rate=(int)90000, encoding-name=(string)H264' ! rtph264depay ! avdec_h264 ! xvimagesink
```

**Reply by svpcom** [reply]
Ping test will not work (will ping interfaces locally) if both ends are on one host.

**Reply by marstears** [reply]
There is no Xwindow on both rx and tx side and I will find a way to do your test.
Thank you.

**Reply by marstears** [reply]
@svpcom 
Do I need write both drone_video and gs_video in wifibroadcast.cfg when insert two cards into same board to do the test?

**Reply by svpcom** [reply]
Yes. You can just use minimal /etc/wifibroadcast.cfg (only add `wifi_channel=XXX` in section `[common]`), because default settings in `telemetry/conf/master.cfg` already contains them.

**Reply by marstears** [reply]
@svpcom 
Hi, I have done your test. Everythin is good when two wifi cards in same tinkerboard. There was no package lost.
![image](https://user-images.githubusercontent.com/28802367/113864024-fa210780-97dc-11eb-8b40-3cd294fe27b2.png)
But I still can not make it work when two cards on two boards.

**Reply by svpcom** [reply]
Check what is difference between linux config on these boards. Try to swap sd-cards between boards to check is this problem software or hardware related

**Reply by marstears** [reply]
@svpcom 
Hi, thank you svpcom. I finally make it work! I
I turned off my office wifi router, and then wifibroadcast seems to be working properly. I don't know why because I'm very sure my router use channel 161 and wifibroadcast use channel 165.
This is the screenshot of server gs log over 20 minutes, there still had same package lost but video was stable.
![image](https://user-images.githubusercontent.com/28802367/113991126-bda8e680-9884-11eb-9b6e-470b65ed7387.png)
![image](https://user-images.githubusercontent.com/28802367/113991415-fe086480-9884-11eb-92ba-086018abbdda.png)

**Was it normal? What else can I do for wifibroadcast optimization?**
